### PR TITLE
simulator: add gamm and tokenfactory messages

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

This is my first contribution to the refactored simulator. The purpose of this PR is to ensure I am heading in the right direction with the added messages. If so, I will continue to add messages for all tested modules.

## Brief Changelog
 
  - Adds a SelAddrWithDenom function (I am sure there is a way to modify the SelAddrWithDenoms function to only take one denom, but did this as a work around)
  - Changes from NewMsgBasedAction to NewCurriedMsgBasedAction
  - Adds RandomExitPoolMsg, RandomCreatePoolMsg, RandomMsgCreateDenom, and RandomMsgMintDenom
  - Changes the default genesis params of tokenfactory to take "stake" denom in place of "uosmo" denom due to current sdk constraints
  - Changes the default genesis params of gamm to take "stake" denom in place of "uosmo" denom for pool creation fee
  - Adds checks to ensure randomly selected wallet has enough stake token to create pool

## Testing and Verifying

This change is already covered by existing tests, such as the simulator itself

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: simulation/types/account.go [hunk 1]
```diff
@@ -95,6 +95,43 @@ func (sim *SimCtx) SelAddrWithDenoms(ctx sdk.Context, denoms []string) (simulati
...
+		return !sim.App.GetBankKeeper().GetBalance(ctx, acc.Address, denom).Amount.IsZero()
+	}
+
+	acc, accExists := sim.RandomSimAccountWithConstraint(accHasDenoms)
+	if !accExists {
+		return acc, sdk.Coin{}, false
+	}
+	balance := sim.RandCoinSubset(ctx, acc.Address, denomSlice)
+	balanceCoin := balance[0]
+	return acc, balanceCoin, true
```
Comment (ValarDragon) [inline simulation/types/account.go:114]:
```suggestion
	acc, subsetCoins, found := sim.SelAddrWithDenoms(ctx, []string{denom}
        if !fouind {
          return acc, sdk.Coin{}, found
        }
        return acc, subsetCoins[0], found
```

File: simulation/types/account.go [hunk 2]
```diff
@@ -95,6 +95,43 @@ func (sim *SimCtx) SelAddrWithDenoms(ctx sdk.Context, denoms []string) (simulati
...
+		}
+		coins = sim.App.GetBankKeeper().SpendableCoins(ctx, acc.Address)
+		return false
+	}
+	acc, found := sim.RandomSimAccountWithConstraint(accHasBal)
+	if !found {
+		return simulation.Account{}, sdk.Coins{}, false
+	}
+	return acc, coins, true
+}
```
Comment (ValarDragon) [inline simulation/types/account.go:126]: I recommend we make this function not return coins, and then just become: 
```
func (sim *SimCtx) RandomSimAccountWithKDenoms(ctx sdk.Context, k int) (simulation.Account, sdk.Coins, bool) {
	accHasBal := func(acc simulation.Account) bool {
		return len(sim.App.GetBankKeeper().SpendableCoins(ctx, acc.Address)) >= k 
	}
	return sim.RandomSimAccountWithConstraint(accHasBal)
}
```

After looking at the API consequences / this in more detail, I think the previous API I made for JoinPool, which combined getting the subset of tokens with getting the account, was a bad design.

File: x/tokenfactory/simulation/sim_msgs.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	// select a pseudo-random denom using alpha-numeric characters
+	var letterRunes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890")
+	// denom can be no greater than 44 characters
+	denomLen := simulation.RandLTBound(sim, 44)
+
+	b := make([]rune, denomLen)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	subdenom := string(b)
```
Comment (ValarDragon) [inline x/tokenfactory/simulation/sim_msgs.go:30]:
```suggestion
	subdenom := sim.RandStringOfLength(44)
```

File: x/tokenfactory/simulation/sim_msgs.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// RandomMsgCreateDenom creates a random tokenfactory denom that is no greater than 44 alphanumeric characters
+func RandomMsgCreateDenom(k keeper.Keeper, sim *simulation.SimCtx, ctx sdk.Context) (*types.MsgCreateDenom, error) {
+	// select a pseudo-random simulation account
+	sender, err := sim.RandomSimAccountWithBalance(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ValarDragon) [inline x/tokenfactory/simulation/sim_msgs.go:19]: Do we need a balance, or can we just do `sim.RandomSimAccount` ?

File: x/tokenfactory/simulation/sim_msgs.go (hunk 3 above)
Comment (czarcas7ic) [reply x/tokenfactory/simulation/sim_msgs.go:30]: A+ for creativity though right

Comment (czarcas7ic) [reply x/tokenfactory/simulation/sim_msgs.go:30]: F for execution

File: simulation/types/account.go (hunk 2 above)
Comment (ValarDragon) [reply simulation/types/account.go:126]: Then we make a second function with thats `GetRandSubsetOfKDenoms(ctx, addr, k)`

File: x/tokenfactory/simulation/sim_msgs.go (hunk 3 above)
Comment (ValarDragon) [reply x/tokenfactory/simulation/sim_msgs.go:30]: haha!

For sure! The way this RandStringOfLength function got made in the SDK was basically me doing the exact same thing you did lolll

**Comment by ValarDragon** [comment]
Nice job on this!!

File: x/gamm/simulation/sim_msgs.go [hunk 5]
```diff
@@ -44,9 +39,57 @@ func RandomJoinPoolMsg(k keeper.Keeper, sim *simulation.SimCtx, ctx sdk.Context)
...
+		Sender:        sender.Address.String(),
+		PoolId:        pool_id,
+		ShareInAmount: gammShares.Amount,
+		TokenOutMins:  tokenOutMins,
+	}, nil
+}
+
+// RandomCreatePoolMsg attempts to find an account with two or more distinct denoms and attempts to send a
+// create pool message composed of those denoms
+func RandomCreatePoolMsg(k keeper.Keeper, sim *simulation.SimCtx, ctx sdk.Context) (*balancertypes.MsgCreateBalancerPool, error) {
```
Comment (ValarDragon) [inline x/gamm/simulation/sim_msgs.go:69]:
```suggestion
func RandomCreateUniV2lMsg(k keeper.Keeper, sim *simulation.SimCtx, ctx sdk.Context) (*balancertypes.MsgCreateBalancerPool, error) {
```

UniV2 = Balancer with two denoms, and 50-50 weight!

File: x/gamm/simulation/sim_msgs.go [hunk 6]
```diff
@@ -44,9 +39,57 @@ func RandomJoinPoolMsg(k keeper.Keeper, sim *simulation.SimCtx, ctx sdk.Context)
...
+	// TODO: pseudo-randomly generate swap and exit fees
+	poolParams := &balancertypes.PoolParams{
+		SwapFee: sdk.NewDecWithPrec(1, 2),
+		ExitFee: sdk.NewDecWithPrec(1, 2),
+	}
+
+	// from the above selected account, determine the token type and respective weight needed to make the pool
+	for i := 0; i < len(poolCoins); i++ {
+		poolAssets = append(poolAssets, balancertypes.PoolAsset{
+			Weight: poolCoins[i].Amount,
```
Comment (ValarDragon) [inline x/gamm/simulation/sim_msgs.go:86]:
```suggestion
			Weight: sdk.OneInt(),
```

(or whatever the API for that is)

File: simulation/types/account.go (hunk 2 above)
Comment (czarcas7ic) [reply simulation/types/account.go]: @ValarDragon was this what you were suggesting? https://github.com/osmosis-labs/osmosis/pull/2091/commits/64c385fad5956521c42778e9ffa7fd10af24f8cd

Side note but still an issue, for some reason when running these tests, no gamm or token factory results are showing in the end report. Did we disable this in one of the merges?

**Comment by mattverse** [comment]
Is this r4r?

**Comment by czarcas7ic** [comment]
added the mint token factory denom logic so now createuniv2pool works as well. The only thing missing I think is restricting accounts that dont have enough for the pool creation fee from creating pools, then I think it is good to go

File: x/gamm/simulation/sim_msgs.go [hunk 7]
```diff
@@ -13,40 +12,138 @@ import (
...
+
+	if tokenIn.Len() != 0 {
+		if !(neededLpLiquidity.DenomsSubsetOf(tokenIn) && tokenIn.IsAllGTE(neededLpLiquidity)) {
+			return &gammtypes.MsgJoinPool{}, fmt.Errorf("TokenInMaxs is less than the needed LP liquidity to this JoinPoolNoSwap,"+
+				" upperbound: %v, needed %v", tokenIn, neededLpLiquidity)
+		} else if !(tokenIn.DenomsSubsetOf(neededLpLiquidity)) {
+			return &gammtypes.MsgJoinPool{}, fmt.Errorf("TokenInMaxs includes tokens that are not part of the target pool,"+
+				" input tokens: %v, pool tokens %v", tokenIn, neededLpLiquidity)
+		}
+	}
```
Comment (ValarDragon) [inline x/gamm/simulation/sim_msgs.go:76]: uhhh, we definitely to reduce this to be far less logic lol

File: x/tokenfactory/simulation/sim_msgs.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+	iterator := store.Iterator(nil, nil)
+	defer iterator.Close()
+
+	denoms := []string{}
+	for ; iterator.Valid(); iterator.Next() {
+		denoms = append(denoms, string(iterator.Key()))
+	}
+	randIndex := simulation.RandLTBound(sim, len(denoms))
+	denom := denoms[randIndex]
```
Comment (ValarDragon) [inline x/tokenfactory/simulation/sim_msgs.go:44]:
```suggestion
	denom := simulation.RandSelect(sim, denoms)
```

**Pull request merged**